### ISSUE TITLE
Back buffer: No drawing during resize of widget

### ISSUE THREAD

**Issue by franzlst** [gaphor/gaphas#63]
Is it intended/desired that during a resize operation of the `GtkView`, the canvas is not drawn, i.e. only the background color is visible?

Our canvas is nested in (several) `Gtk.Paned` and resides next to a `Gtk.Revealer`. Thus, in three situations, no state machine now shown:

* the panes are resized
* the revealer slides up
* the window is resized

Personally, I would want the canvas to be drawn in all these situations.

**Comment by amolenaar** [comment]
That's odd. In the demo app it all works fine, inside a `Gtk.ScrolledWindow`. What would be the easiest way I can reproduce this?

Note: I've been testing on macOS

**Comment by franzlst** [comment]
I can reproduce this using the `demo.py`. Resizing the window (quickly) hides all elements. 

I cannot say what the reason is. I added `print` statements in both `update_back_buffer` and `do_draw`. When resizing, they are called in alternating order:

```
update_back_buffer for rectangle 721 698
do draw
update_back_buffer for rectangle 794 722
do draw
update_back_buffer for rectangle 855 741
do draw
update_back_buffer for rectangle 916 757
do draw
update_back_buffer for rectangle 981 773
do draw
```

I am using Linux (OSL 42.3) and pycairo 1.18.

**Comment by amolenaar** [comment]
I checked with Ubuntu in Virtualbox, but I can not reproduce this. I notice some flickering when I resize the window, but it will always paint.

**Comment by amolenaar** [comment]
Can you check if either `do_size_allocate` and `do_configure_event` are called as well? If the later is not called, nothing will be rendered.

**Comment by franzlst** [comment]
I think I found the reasen. I added another print statement in `do_configure_event`. When resizing, the order of signal/callbacks seems to be:
1. `update_back_buffer` (creates the back buffer)
2. `do_configure_event` (__resets the back buffer__)
3. `on_draw` (draws from the (resetted) buffer)

**Comment by amolenaar** [comment]
The update is coming from do_size_allocate(). I noticed this method is called before `do_configure_event`. My PR drops size_allocate and makes every update go through `do_configure_event`. I expect that would fix it then.

**Comment by franzlst** [comment]
No it doesn't help.

What _does_ help is removing the two `AsyncIO` decorator from `GtkView.update` _and_ `GtkView.update_back_buffer`. Then, resizing does not cause any problems. With the decorators removed, this is the stack of `update_back_buffer`:

```
File "/home_local/stei_fn/gaphas/examples/demo.py", line 495, in <module>
    main()
  File "/home_local/stei_fn/gaphas/examples/demo.py", line 477, in main
    Gtk.main()
  File "/home_local/stei_fn/gaphas/gaphas/view.py", line 861, in do_configure_event
    self.update()
  File "/home_local/stei_fn/gaphas/gaphas/view.py", line 759, in update
    self.update_back_buffer()
  File "/home_local/stei_fn/gaphas/gaphas/view.py", line 784, in update_back_buffer
    traceback.print_stack()
```

and the order of calls is
1. configure
2. update_back_buffer
3. draw

With at least one decorator in place, the stack looks like this:
```
  File "/home_local/stei_fn/gaphas/examples/demo.py", line 495, in <module>
    main()
  File "/home_local/stei_fn/gaphas/examples/demo.py", line 477, in main
    Gtk.main()
  File "/home_local/stei_fn/gaphas/gaphas/decorators.py", line 144, in async_wrapper
    func(*args, **kwargs)
  File "/home_local/stei_fn/gaphas/gaphas/view.py", line 784, in update_back_buffer
    traceback.print_stack()
```

So apparently, the decorator moves the `update_back_buffer` after `draw`.

**Comment by amolenaar** [comment]
Okay... I would have expected another `draw` to be scheduled, since `update_back_buffer` calls `super(GtkView, self).queue_draw_area()`. That should cause the widget to be redrawn. At least, that's what I would expect.

The async decorator on GtkView.update has been there for a long time. It reduces the view updates if a couple of canvas updates happen sequentially. The async decorator on `update_back_buffer` we can drop, although I'm curious why no draw is scheduled for the window.

I changed the call to `queue_draw_area` to a refresh of the complete window. Wonder if that helps...

**Comment by danyeaw** [comment]
Closed by #65 :tada:

**Comment by franzlst** [comment]
Thanks for fixing!
